src/ServiceControl/CompositeViews/Messages/MessageViewQueryExtensions.cs

```diff
@@ -0,0 +1,216 @@
+                var totalCount = responseMessage.Headers.GetValues("Total-Count").Select(int.Parse).Cast<int?>().FirstOrDefault() ?? -1;
+                IEnumerable<string> etags;
+                string etag = null;
+                if (responseMessage.Headers.TryGetValues("ETag", out etags))
+                {
+                    etag = etags.ElementAt(0);
+                }
+
+                IEnumerable<string> lastModifiedValues;
+                var lastModified = DateTime.UtcNow;
+                if (responseMessage.Headers.TryGetValues("Last-Modified", out lastModifiedValues))
+                {
+                    lastModified = DateTime.ParseExact(lastModifiedValues.ElementAt(0), "R", CultureInfo.InvariantCulture);
+                }
+
+                return new PartialQueryResult
+                {
+                    Messages = messages,
+                    Header = new HeaderInfo(etag, lastModified, totalCount)
+                };
+            }
+        }
+
+        static MessagesView[] SortResults(DynamicDictionary queryString, ICollection<MessagesView> combinedMessages)
+        {
+            //Set the default sort to time_sent if not already set
+            string sortBy = queryString.ContainsKey("sort")
+                ? queryString["sort"]
+                : "time_sent";
+
+            //Set the default sort direction to `desc` if not already set
+            string sortOrder = queryString.ContainsKey("direction")
+                ? queryString["direction"]
+                : "desc";
+
+            Func<MessagesView, IComparable> keySelector = m => m.TimeSent;
+            switch (sortBy)
+            {
+                case "id":
+                case "message_id":
+                    keySelector = m => m.MessageId;
+                    break;
+
+                case "message_type":
+                    keySelector = m => m.MessageType;
+                    break;
+
+                case "critical_time":
+                    keySelector = m => m.CriticalTime;
+                    break;
+
+                case "delivery_time":
+                    keySelector = m => m.DeliveryTime;
+                    break;
+
+                case "processing_time":
+                    keySelector = m => m.ProcessingTime;
+                    break;
+
+                case "processed_at":
+                    keySelector = m => m.ProcessedAt;
+                    break;
+
+                case "status":
+                    keySelector = m => m.Status;
+                    break;
+            }
+
+            return sortOrder == "asc"
+                ? combinedMessages.OrderBy(keySelector).ToArray()
+                : combinedMessages.OrderByDescending(keySelector).ToArray();
+        }
+
+        class PartialQueryResult
+        {
+            public IList<MessagesView> Messages { get; set; }
+            public HeaderInfo Header { get; set; }
+        }
+
+        struct HeaderInfo
```

Comment:
If this were QueryStatsInfo or something then it could be used by local as well which just means less params to pass around. You hand results and stats to CombineWithRemote call.

You could even put some of the aggregation logic inside here to clean this up:

```
var results = localResults;
var stats = localStats;
foreach(resultSet)
{
  results.AppendAll(remote.Results);
  stats += remote.Stats;
}
```

All of this:

```
                .WithPagingLinksAndTotalCount(headerInfo.TotalCount, numberOfInstances, currentRequest)
                .WithDeterministicEtag(headerInfo.ETag)
                .WithLastModified(headerInfo.LastModified);
```

Can be replaced with a: `.WithStats(stats)` call